src/daft-io/src/glob.rs

```diff
@@ -0,0 +1,194 @@
+use std::{collections::HashSet, sync::Arc};
+
+use globset::GlobMatcher;
+use lazy_static::lazy_static;
+
+lazy_static! {
+    /// Check if a given char is considered a special glob character
+    /// NOTE: we use the `globset` crate which defines the following glob behavior:
+    /// https://docs.rs/globset/latest/globset/index.html#syntax
+    static ref GLOB_SPECIAL_CHARACTERS: HashSet<char> = HashSet::from(['*', '?', '{', '}', '[', ']']);
+}
+
+const SCHEME_SUFFIX_LEN: usize = "://".len();
+
+pub(crate) struct GlobState {
+    // Current path in dirtree and glob_fragments
+    pub current_path: String,
+    pub current_fragment_idx: usize,
+
+    // Whether we have encountered wildcards yet in the process of parsing
+    pub wildcard_mode: bool,
+
+    // Carry along expensive data as Arcs to avoid recomputation
+    pub glob_fragments: Arc<Vec<GlobFragment>>,
+    pub full_glob_matcher: Arc<GlobMatcher>,
+}
+
+impl GlobState {
+    pub fn current_glob_fragment(&self) -> &GlobFragment {
+        &self.glob_fragments[self.current_fragment_idx]
+    }
+
+    pub fn advance(&self, path: String, idx: usize) -> Self {
```

Comment:
Ok, I left it as `fn_(&self)` though because it was a little annoying to have to have the caller `.clone()` (the caller potentially needs to call `.advance` on the same object many times slightly differently each time, and it's easier to do it on a &self).